[DI - Step2] 러너덕 미션 제출합니다.

Deocksoo opened the pull request:
안녕하세요 후니, 리뷰이 러너덕입니다 :)
step2 요구사항들을 구현하였고, 더불어 지난 step1때 주셨던 피드백도 모두 반영해서 리펙토링 하였습니다. 이번 미션이 다소 까다로워서 생각보다 시간이 많이 걸렸어요.
현재 BeanFactory의 createBean 메소드에서 configuration에 의해 생성되는 bean과 component scan에 의해 생성되는 bean들을 모두 생성하여주고 있는데요, component scan으로 scan한 class들을 bean으로 만들 때에는 configuration에 null이 들어가도록 만들어둔 상태입니다. 이 부분이 상당히 마음에 걸리는데, 좀처럼 다른 좋은 해결 방안이 떠오르지 않네요. 리뷰어님의 의견은 어떠신지 궁금합니다.
이 부분 외에도 수정해야할 부분이 군데군데 꽤 있을것 같아요, 부족한 부분들이 있다면 피드백 부탁드립니다 :)
감사합니다!

pkch93 committed:
refactor: BeanFactory 클래스 중복제거

- beans.containKey 중복 제거

pkch93 committed:
docs: step-2 요구사항 정리

pkch93 committed:
feat: Configuration Scan 및 ComponentScan 기능 구현

pkch93 committed:
docs: 의문사항 정리

pkch93 committed:
feat: Configuration 클래스 스캔 추가

pkch93 committed:
feat: BeanDefinition 구현

* 두 가지 방식의 Bean 생성 전략을 추상화하여 BeanCreator를 만들었다.

pkch93 committed:
feat: BeanFactoryUtils.findBeanDefinition 추가

pkch93 committed:
feat: BeanDefinitionFactory 구현

pkch93 committed:
refactor: BeanFactory에 BeanDefinitionFactory 적용

Deocksoo committed:
refactor: 불필요한 logging 제거

Deocksoo committed:
refactor: BeanFactoryUtils 하드코딩된 숫자 상수처리

Deocksoo committed:
fix: BeanScannerTest 실패하는 테스트 수정

Deocksoo committed:
refactor: AnnotationHandlerMapping 메서드 가독성 개선

Deocksoo committed:
refactor: BeanDefinitionFactory 잘못된 변수명 수정 beanCreation -> beanCreator

Deocksoo committed:
refactor: BeanDefinitionFactory 코드 정리

각 메서드가 하나의 일만 하도록 수정하고, 불필요한 메서드 인자들을 제거합니다.

Deocksoo committed:
docs: 의문사항 제거

Deocksoo committed:
refactor: BeanFactoryUtils의 getInjectedConstructor에서 null 반환하지 않도록 수정

Deocksoo committed:
refactor: AnnotationHandlerMapping에서 @RequestMapping scan하는 기능 분리

Deocksoo committed:
refactor: BeanFactory 코드 정리

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanDefinitionFactory.java:
```diff
@@ -0,0 +1,93 @@
+
+        return new BeanDefinition(
+                clazz,
+                null,
```
질문주셨던 내용 중에 해당 부분이 고민이신 것 같네요!
BeanDefinition을 추상클래스와 인터페이스를 사용해서 분리하면 불필요한 부분을 없앨 수 있지 않을까요? (공통적인 부분만 두어서)

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanScanner.java:
```diff
@@ -10,14 +12,36 @@
     private static final Logger log = LoggerFactory.getLogger(BeanScanner.class);
 
+    @SuppressWarnings("unchecked")
+    public static Set<Class<?>> scanConfiguration(String basePackage) {
```
메서드가 너무 많은 일을 하는 것 같아요. 분리하면 조금 더 깔끔할 것 같아요!

Deocksoo replied:
앗 이 부분도 리펙토링이 필요한 부분이네요! 분리하도록 하겠습니다 :)

jeonghoon1107 commented on slipp/src/main/java/slipp/SlippWebApplicationInitializer.java:
```diff
@@ -19,7 +20,8 @@
     @Override
     public void onStartup(ServletContext servletContext) throws ServletException {
-        BeanFactory beanFactory = new BeanFactory(BeanScanner.scan("slipp"));
+        BeanDefinitionFactory beanDefinitionFactory = new BeanDefinitionFactory(BeanScanner.scanConfiguration("slipp"));
```
webServerLauncher.class의 basePackage를 주어도 될 것 같아요!

Deocksoo replied:
basePackage를 주는 것이 좀더 낫겠네요, 수정하겠습니다!

jeonghoon1107 reviewed (changes requested):
안녕하세요 러너덕!
깔끔하게 구현 잘 하셨네요 👍 
상당히 마음에 걸리는 부분이 있다고 하셔서 리팩토링을 위해 바로 머지 하지 않겠습니다!
질문 주신 것에 대한 답변은 해당 하는 코드 부분에 달아놓았으니 확인부탁드려요 :)

Deocksoo committed:
refactor: BeanScanner.scanConfiguration 메서드 분리

한 메서드에서 너무 많은 일을 수행하고 있어, 이를 분리하였습니다.

Deocksoo committed:
refactor: SlippWebApplicationInitializer에서 하드코디된 basePackage 이름 변경

WebServerLauncher 클래스가 선언된 패키지를 basePackage로 하도록 수정합니다.

Deocksoo committed:
refacotor: BeanDefinition을 추상클래스와 인터페이스를 사용하여 분리

@Component 어노테이션을 통해 생성되는 bean과 @Configuration 클래스로부터 생성되는
bean을 구분하기 위해 BeanDefinition을 추상클래스와 인터페이스를 사용해 분리하였습니다.

Deocksoo commented:
안녕하세요 후니, 리뷰이 러너덕입니다.
레벨 4 프로젝트와 병행하느라 DI미션 제출이 많이 늦어져버렸네요ㅠㅠ

지난 리뷰때 주셨던 피드백을 모두 반영해보는 것이 좋을것 같아서 말씀해주신 부분들 수정해서 요청 보내드립니다. 말씀해주신 대로 BeanDefinition을 추상 클래스와 인터페이스로 분리해보는 것이 좋은 시도일 것 같아서 분리해보았는데요, 적절하게 잘 분리되었는지 확인 부탁드려요 :)

늦게까지 신경써주져서 감사합니다!

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/AbstractBeanDefinition.java:
```diff
@@ -0,0 +1,35 @@
+public abstract class AbstractBeanDefinition implements BeanDefinition {
+    private Class<?> type;
+    private BeanCreator beanCreator;
+    private List<Class<?>> parameters;
```
일급 콜렉션을 사용해보면 어떨까요?

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanScanner.java:
```diff
@@ -10,14 +12,38 @@
+    private static Set<Class<?>> scanComponentsInBasePackagesOf(Set<Class<?>> configurations) {
+        Set<Class<?>> components = Sets.newHashSet();
+
+        configurations.stream()
```
체이닝이 너무 기네요..
분리해보는 것은 어떨까요?

jeonghoon1107 reviewed (approved):
안녕하세요 러너덕!
오랜만의 pr이라 기억하기 쉽지 않았을텐데 깔끔하게 리팩토링 하셨네요 :)
고생 많으셨습니다!
